pkg/apply/solver/solver.go

```diff
@@ -86,74 +87,139 @@ func (tq *TaskQueue) ToActionGroups() []event.ActionGroup {
-// Returns a pointer to the Builder to chain function calls.
-func (t *TaskQueueBuilder) AppendInvAddTask(inv inventory.Info, applyObjs object.UnstructuredSet,
-	dryRun common.DryRunStrategy) *TaskQueueBuilder {
-	applyObjs = t.Collector.FilterInvalidObjects(applyObjs)
-	klog.V(2).Infoln("adding inventory add task (%d objects)", len(applyObjs))
-	t.tasks = append(t.tasks, &task.InvAddTask{
-		TaskName:  fmt.Sprintf("inventory-add-%d", t.invAddCounter),
-		InvClient: t.InvClient,
-		InvInfo:   inv,
-		Objects:   applyObjs,
-		DryRun:    dryRun,
-	})
-	t.invAddCounter++
+// WithApplyObjects sets the apply objects and returns the builder for chaining.
+func (t *TaskQueueBuilder) WithApplyObjects(applyObjs object.UnstructuredSet) *TaskQueueBuilder {
+	t.applyObjs = applyObjs
 	return t
 }
 
-// AppendInvSetTask appends an inventory set task to the task queue.
-// Returns a pointer to the Builder to chain function calls.
-func (t *TaskQueueBuilder) AppendInvSetTask(inv inventory.Info, dryRun common.DryRunStrategy) *TaskQueueBuilder {
-	klog.V(2).Infoln("adding inventory set task")
-	prevInvIds, _ := t.InvClient.GetClusterObjs(inv)
-	t.tasks = append(t.tasks, &task.InvSetTask{
-		TaskName:      fmt.Sprintf("inventory-set-%d", t.invSetCounter),
-		InvClient:     t.InvClient,
-		InvInfo:       inv,
-		PrevInventory: prevInvIds,
-		DryRun:        dryRun,
-	})
-	t.invSetCounter++
+// WithPruneObjects sets the prune objects and returns the builder for chaining.
+func (t *TaskQueueBuilder) WithPruneObjects(pruneObjs object.UnstructuredSet) *TaskQueueBuilder {
+	t.pruneObjs = pruneObjs
 	return t
 }
 
-// AppendDeleteInvTask appends to the task queue a task to delete the inventory object.
-// Returns a pointer to the Builder to chain function calls.
-func (t *TaskQueueBuilder) AppendDeleteInvTask(inv inventory.Info, dryRun common.DryRunStrategy) *TaskQueueBuilder {
-	klog.V(2).Infoln("adding delete inventory task")
-	t.tasks = append(t.tasks, &task.DeleteInvTask{
-		TaskName:  fmt.Sprintf("delete-inventory-%d", t.deleteInvCounter),
-		InvClient: t.InvClient,
-		InvInfo:   inv,
-		DryRun:    dryRun,
-	})
-	t.deleteInvCounter++
-	return t
+// Build returns the queue of tasks that have been created
+func (t *TaskQueueBuilder) Build(o Options) *TaskQueue {
+	var tasks []taskrunner.Task
+
+	// reset counters
+	t.applyCounter = 0
+	t.pruneCounter = 0
+	t.waitCounter = 0
+
+	applyObjs := t.Collector.FilterInvalidObjects(t.applyObjs)
+	pruneObjs := t.Collector.FilterInvalidObjects(t.pruneObjs)
+
+	if len(applyObjs) > 0 {
+		klog.V(2).Infoln("adding inventory add task (%d objects)", len(applyObjs))
+		tasks = append(tasks, &task.InvAddTask{
+			TaskName:  "inventory-add-0",
+			InvClient: t.InvClient,
+			InvInfo:   t.invInfo,
+			Objects:   applyObjs,
+			DryRun:    o.DryRunStrategy,
+		})
+	}
+
+	// Use the "depends-on" annotation to create a graph, ands sort the
+	// objects to apply into sets using a topological sort.
+	if len(applyObjs) > 0 {
```

Comment:
This **if** statement looks the same as the one directly above; can this statement be removed and the code consolidated into the above block?

Reply:
i tried to keep the tasks logically separated for readability, but yes, i could "reduce" the code a bit.

Reply:
done